crates/ruff_python_semantic/src/red_knot/semantic_index/symbol.rs

```diff
@@ -0,0 +1,360 @@
+// Allow unused underscore violations generated by the salsa macro
+// TODO(micha): Contribute fix upstream
+#![allow(clippy::used_underscore_binding)]
+
+use std::hash::{Hash, Hasher};
+use std::ops::{Index, Range};
+
+use bitflags::bitflags;
+use hashbrown::hash_map::RawEntryMut;
+use rustc_hash::FxHasher;
+use smallvec::{smallvec, SmallVec};
+
+use crate::Db;
+use ruff_db::vfs::VfsFile;
+use ruff_index::{newtype_index, IndexVec};
+
+use crate::name::Name;
+use crate::red_knot::semantic_index::definition::Definition;
+use crate::red_knot::semantic_index::{symbol_table, SymbolMap};
+
+#[derive(Eq, PartialEq, Debug)]
+pub struct Symbol {
+    name: Name,
+    flags: SymbolFlags,
+    scope: ScopeId,
+
+    /// The nodes that define this symbol, in source order.
+    definitions: SmallVec<[Definition; 4]>,
+}
+
+impl Symbol {
+    fn new(name: Name, scope: ScopeId, definition: Option<Definition>) -> Self {
+        Self {
+            name,
+            scope,
+            flags: SymbolFlags::empty(),
+            definitions: smallvec![definition.unwrap_or(Definition::Unbound)],
```

Comment:
@carljm I copied this over from the existing semanitc model. Always pushing `Unbound` feels off. Could we insert unbound in the read path instead of when creating the symbols?

Reply:
By "existing semantic model" here I assume you mean the existing (non-red-knot) ruff semantic model? Because the existing non-salsa red-knot semantic model doesn't do this, it already inserts Unbound in the read path (in ReachableDefinitionsIterator.)